### DISCUSSION TITLE
Material-UI without SSR

### DISCUSSION THREAD

**Discussion by underfisk** [vercel/next.js#22006] [Help]
Is there anyone using MUI without SSR? For development its fine but my static web app does not use SSR in production, it just uses locally because of MUI but i wanted to opt-out and couldn't find any official supported method. to do so but there might be some work around to it

**Comment by mkrtchian** [comment]
You can take a look at the [Next.js example on Material UI repository](https://github.com/mui-org/material-ui/tree/master/examples/nextjs). It works without SSR, you can even `next export` that project.

**Reply by underfisk** [reply]
@mkrtchian The issue is that it has getInitialProps in document which makes nextjs not fully export in static, that's what i was talking about because i get warnings saying that i've opted out SSG or static due to the fact it is using server-side stylings

**Reply by mkrtchian** [reply]
If you try the example from Material UI repo, you will get no warning during build, and the pages will be static:
```bash
curl https://codeload.github.com/mui-org/material-ui/tar.gz/master | tar -xz --strip=2  material-ui-master/examples/nextjs
cd nextjs
npm install
npm run build
```

```
> next build

info  - Creating an optimized production build  
info  - Compiled successfully
info  - Collecting page data  
info  - Generating static pages (3/3)
info  - Finalizing page optimization  

Page                                                           Size     First Load JS
┌ ○ /                                                          418 B          92.6 kB
├   /_app                                                      0 B            85.2 kB
├ ○ /404                                                       3.03 kB        88.2 kB
└ ○ /about                                                     5.54 kB        97.7 kB
+ First Load JS shared by all                                  85.2 kB
  ├ chunks/701e55b9450706e1e000a5c1b3746a90803fda45.0e26a9.js  19.5 kB
  ├ chunks/commons.8b6958.js                                   13.1 kB
  ├ chunks/framework.b016d4.js                                 42.1 kB
  ├ chunks/main.ab3cf6.js                                      6.62 kB
  ├ chunks/pages/_app.6b2375.js                                3.2 kB
  └ chunks/webpack.50bee0.js                                   751 B

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
   (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
```

But if you add the following code in `pages/_app.js`:
```javascript
MyApp.getInitialProps = async () => {};
```

You will get the warning and the λ sign next to the pages signifying that they will be server rendered:
```
> next build

info  - Creating an optimized production build  
info  - Compiled successfully
Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps`
Read more: https://err.sh/next.js/opt-out-auto-static-optimization

Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps`
Read more: https://err.sh/next.js/opt-out-auto-static-optimization

info  - Collecting page data  
Page                                                           Size     First Load JS
┌ λ /                                                          418 B          92.8 kB
├   /_app                                                      0 B            85.4 kB
├ λ /404                                                       3.03 kB        88.5 kB
└ λ /about                                                     5.54 kB        97.9 kB
+ First Load JS shared by all                                  85.4 kB
  ├ chunks/701e55b9450706e1e000a5c1b3746a90803fda45.0e26a9.js  19.5 kB
  ├ chunks/commons.8b6958.js                                   13.1 kB
  ├ chunks/framework.b016d4.js                                 42.1 kB
  ├ chunks/main.ab3cf6.js                                      6.62 kB
  ├ chunks/pages/_app.8d08af.js                                3.43 kB
  └ chunks/webpack.50bee0.js                                   751 B

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
   (ISR)     incremental static regeneration (uses revalidate in getStaticProps)
```

There is [an interesting comment on the code](https://github.com/mui-org/material-ui/blob/master/examples/nextjs/pages/_document.js#L27):
> // `getInitialProps` belongs to `_document` (instead of `_app`),
> // it's compatible with server-side generation (SSG).

And here is [another comment resource going in the same direction](https://github.com/vercel/next.js/discussions/20482#discussioncomment-252823).

And by the way I have myself already created a project using Material UI with `getInitialProps` in `pages/_document.js`, and I have been able to `next export` it and serve it without any node.js server, as fully static resource.

**Reply by underfisk** [reply]
@mkrtchian I see, i seems a false positive then because i'm got a running application in prod with the same config, i can export too but since it always says its not optimized i was looking for an alternative

**Reply by mkrtchian** [reply]
To be sure if your pages are static or use SSR, you could look at the output of `next build`. For each page:
- **λ**: it uses SSR.
- **○**: it is static.
- **●**: it uses SSG (getStaticProps), so here also static.

If you have a λ on all your pages that do not have `getStaticProps`, then the warning may be real.

Even if you have no `getInitialProps` in the `pages/_app.js`, one of the libraries you use may have put it there for you, like in [#12476](https://github.com/vercel/next.js/issues/12476) or [#14537](https://github.com/vercel/next.js/discussions/14537).
